### PR TITLE
fix(cli-repl): json-clone objects before passing them to pino MONGOSH…

### PR DESCRIPTION
**Pull request by addaleax**
…-884

### CODE REVIEW

File: packages/cli-repl/src/setup-logger-and-telemetry.ts [hunk 1]
```diff
@@ -355,7 +372,10 @@ export default function setupLoggerAndTelemetry(
   });
 
   bus.on('mongosh-sp:connect-heartbeat-failure', function(ev: SpConnectHeartbeatFailureEvent) {
-    log.info('mongosh-sp:connect-heartbeat-failure', ev);
+    log.info('mongosh-sp:connect-heartbeat-failure', {
+      ...ev,
+      failure: ev.failure?.message
+    });
```
Comment (addaleax) [inline packages/cli-repl/src/setup-logger-and-telemetry.ts:378]: This is more of a drive-by fix, it matches what we do below. Otherwise, `Error` instances are always just serialized to `{}` (regardless of the rest of this patch).

File: packages/cli-repl/src/setup-logger-and-telemetry.ts [hunk 2]
```diff
@@ -186,11 +203,11 @@ export default function setupLoggerAndTelemetry(
   });
 
   bus.on('mongosh:setCtx', function(args: ApiEvent) {
-    log.info('mongosh:setCtx', args);
+    log.info('mongosh:setCtx', jsonClone(args));
   });
 
   bus.on('mongosh:api-call', function(args: ApiEvent) {
-    log.info('mongosh:api-call', redactInfo(args));
+    log.info('mongosh:api-call', redactInfo(jsonClone(args)));
```
Comment (mcasimir) [inline packages/cli-repl/src/setup-logger-and-telemetry.ts:210]: is it the other way around? or `redactInfo` also modifies the object?

**Review by mcasimir — APPROVED** [review]

**Review by rose-m — APPROVED** [review]

File: packages/cli-repl/src/setup-logger-and-telemetry.ts (hunk 2 above)
Comment (addaleax) [reply packages/cli-repl/src/setup-logger-and-telemetry.ts:210]: @mcasimir `redactInfo` does not modify its input, so the order does not matter here

Comment (mcasimir) [reply packages/cli-repl/src/setup-logger-and-telemetry.ts:210]: Ah ok! What happens if `jsonClone` gets the `_inspected` version?

Comment (addaleax) [reply packages/cli-repl/src/setup-logger-and-telemetry.ts:210]: `redactInfo` works on strings – if it gets an object, it traverses that object and looks for strings to replace. So I think this way it’s fine, or probably even better, because this way `redactInfo` will handle circular input properly as well, which it didn’t before:

```
> c = {}; c.c = c; require('mongodb-redact')(c)
Uncaught RangeError: Maximum call stack size exceeded
    at Object.toString (<anonymous>)
    at objectToString (/home/addaleax/src/mongosh/packages/cli-repl/node_modules/lodash/lodash.js:6598:35)
    at baseGetTag (/home/addaleax/src/mongosh/packages/cli-repl/node_modules/lodash/lodash.js:3111:11)
    at Function.isPlainObject (/home/addaleax/src/mongosh/packages/cli-repl/node_modules/lodash/lodash.js:12105:35)
    at redact (/home/addaleax/src/mongosh/packages/cli-repl/node_modules/mongodb-redact/lib/index.js:5:9)
    at /home/addaleax/src/mongosh/packages/cli-repl/node_modules/lodash/lodash.js:13469:38
```
